### PR TITLE
Add Java11 support for snapshot/restore

### PR DESCRIPTION
**Pull request by tajila**
Add Java11 support for snapshot/restore

Add support for persisting the module and packages. Modify the restore
sequence to not initialize classes that aren't used in Java11

Signed-off-by: Tobi Ajila <atobia@ca.ibm.com>

### CODE REVIEW

**Comment by DanHeidinga** [comment]
I'm planning to review this PR this afternoon.  Thanks Tobi!

File: runtime/vm/jvminit.c [hunk 1]
```diff
@@ -2359,7 +2359,14 @@ IDATA VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved) {
 			}
 
 			if (J2SE_VERSION(vm) >= J2SE_V11) {
-				vm->modularityPool = pool_new(OMR_MAX(sizeof(J9Package),sizeof(J9Module)),  0, 0, 0, J9_GET_CALLSITE(), J9MEM_CATEGORY_MODULES, POOL_FOR_PORT(vm->portLibrary));
+#if defined(J9VM_OPT_SNAPSHOTS)
+				if (IS_SNAPSHOT_RUN(vm)) {
+					vm->modularityPool = pool_new(OMR_MAX(sizeof(J9Package), sizeof(J9Module)),  0, 0, 0, J9_GET_CALLSITE(), J9MEM_CATEGORY_MODULES, POOL_FOR_PORT(VMSNAPSHOTIMPL_OMRPORT_FROM_JAVAVM(vm)));
+				} else if (!IS_RESTORE_RUN(vm))
```
Comment (DanHeidinga) [inline runtime/vm/jvminit.c:2365]:
```suggestion
				if (IS_SNAPSHOT_RUN(vm)) {
					vm->modularityPool = pool_new(OMR_MAX(sizeof(J9Package), sizeof(J9Module)),  0, 0, 0, J9_GET_CALLSITE(), J9MEM_CATEGORY_MODULES, POOL_FOR_PORT(VMSNAPSHOTIMPL_OMRPORT_FROM_JAVAVM(vm)));
				} else if (IS_RESTORE_RUN(vm)) {
					/* vm->modularityPool is already initialized as part of the restore */
				} else
```

Make the 3 cases {snapshot, restore, normal jvm} explicit.

File: runtime/jcl/common/stdinit.c [hunk 2]
```diff
@@ -169,6 +173,12 @@ restoreRunStandardInit(J9JavaVM *vm)
 #endif /* !J9VM_INTERP_MINIMAL_JCL */
 #endif /* J9VM_OPT_SIDECAR */
 
+#if JAVA_SPEC_VERSION >= 11
+	vmFuncs->internalAcquireVMAccess(vmThread);
+	vmFuncs->sendInitEncodings(vmThread);
```
Comment (DanHeidinga) [inline runtime/jcl/common/stdinit.c:178]: Why is this required?  It inits the static fields {`platformEncoding`, `fileEncoding`, `osEncoding`} in System.java.  Aren't they saved in an initialized state?  Why do we reset them on a restore run?

File: runtime/jcl/common/stdinit.c [hunk 3]
```diff
@@ -161,6 +162,9 @@ restoreRunStandardInit(J9JavaVM *vm)
 	}
 	((J9NativeLibrary*)nativeLibrary)->flags |= J9NATIVELIB_FLAG_ALLOW_INL;
 
+	/* Run the JCL initialization code (what used to be JNI_OnLoad) */
+	nativeLibrary->send_lifecycle_event(vmThread, nativeLibrary, "JCL_OnLoad", JNI_VERSION_1_1);
```
Comment (DanHeidinga) [inline runtime/jcl/common/stdinit.c:166]: The other caller of this API looks like:
```
/* Run the JCL initialization code (what used to be JNI_OnLoad) */
	jniVersion = nativeLibrary->send_lifecycle_event(vmThread, nativeLibrary, "JCL_OnLoad", JNI_VERSION_1_1);
	if (!vmFuncs->jniVersionIsValid(jniVersion)) {
		return 1;
	}
```
Should this code validate the JNI version as well?

File: runtime/jcl/common/stdinit.c (hunk 2 above)
Comment (tajila) [reply runtime/jcl/common/stdinit.c:178]: The JCL natives set a native static when encodings aare initialized, this needs to be re-init in the restore run

File: runtime/jcl/common/stdinit.c (hunk 3 above)
Comment (tajila) [reply runtime/jcl/common/stdinit.c]: yes, I've added the JNI version validation

File: runtime/jcl/common/stdinit.c (hunk 2 above)
Comment (DanHeidinga) [reply runtime/jcl/common/stdinit.c:181]: Thanks.  Following that path, it looks like this is the native that needs to be called again:
https://github.com/eclipse/openj9/blob/0aec295571c6c25927af41a629bc915c2129642e/runtime/jcl/common/system.c#L425-L429

Comment (DanHeidinga) [reply runtime/jcl/common/stdinit.c:181]: Can you add this as a comment to the code?  Something like:
```
/* Need to resend System.initEncodings() because it calls into native code which
 * has to reinitialize the native static value done by the "InitializeEncoding" native.
 * See the code in system.c:: getEncoding
 */

**Comment by DanHeidinga** [comment]
minor request to add a comment and then this looks good to me.

**Review by DanHeidinga — APPROVED** [review]
lgtm

**Comment by DanHeidinga** [comment]
Building this locally as the travis build is incredibly slow to launch
